lib/srv/db/server.go

```diff
@@ -811,6 +811,23 @@ func (s *Server) handleConnection(ctx context.Context, clientConn net.Conn) erro
 
 	err = engine.HandleConnection(ctx, sessionCtx)
 	if err != nil {
+		connectionDiagnosticID := sessionCtx.Identity.ConnectionDiagnosticID
+		if connectionDiagnosticID != "" && trace.IsAccessDenied(err) {
+			_, diagErr := s.cfg.AuthClient.AppendDiagnosticTrace(ctx,
+				connectionDiagnosticID,
+				&types.ConnectionDiagnosticTrace{
+					Type:    types.ConnectionDiagnosticTrace_RBAC_DATABASE_LOGIN,
+					Status:  types.ConnectionDiagnosticTrace_FAILED,
+					Details: "Access denied when accessing Database. Ensure your role allows you to use the requested Database Name and Database User.",
```

Comment:
I agree with @greedy52 here, would be nice to differentiate between Teleport RBAC and RDS IAM errors for example. How does this currently behave if you try to connect to an RDS instance and get an IAM error?

Reply:
I focused mostly on self-hosted Postgres
I'll do some tests using in AWS RDS

Reply:
If I remove the `rds-db:connect` from the EC2 Role's policy (where the DB Agent is running) it returns the following error:

I extracted the error message to increase readability:
```
    {
      "traceType": "RBAC_DATABASE_LOGIN",
      "status": "FAILED",
      "details": "Access denied when accessing Database. Ensure your role allows you to use the requested Database Name and Database User.",
      "error": "<see below>"
    },
```
With the following error:
```
"error": "Could not connect to database:

  FATAL: PAM authentication failed for user \"postgres\" (SQLSTATE 28000)

Make sure that Postgres user \"postgres\" has \"rds_iam\" role and Teleport database
agent's IAM policy has \"rds-connect\" permissions (note that IAM changes may
take a few minutes to propagate):

{
        \"Version\": \"2012-10-17\",
    \"Statement\": [
            {
                \"Effect\": \"Allow\",
            \"Action\": \"rds-db:connect\",
            \"Resource\": \"arn:aws:rds-db:us-east-1:278576220453:dbuser:db-QUAJVCUEOKUNSH3O337FFPXEWI/*\"
        }
    ]
}
"
```

The error here is already converted from the original error to add extra info the error using the function mentioned by @greedy52 
[`ConvertConnectError`](https://github.com/gravitational/teleport/blob/65e35120a1e4d356c400e780e22517f358e07fcd/lib/srv/db/common/errors.go#L115-L148).

This will be sent to the UI within the `error` field.

Do you think sending this info is enough?

I'll change the `details` message to be more generic and defer the remaining to the `error` field.


---
There's an issue here because this trace is appended in the DB Agent and then the Proxy itself adds another trace with failed state. There should be only one error trace, so I'll remove the one in the Proxy.